events/api.py

```diff
@@ -1950,6 +1946,64 @@ class Meta:
         exclude = ["id", "event"]
 
 
+# RegistrationSerializer is in this file to avoid circular imports
+class RegistrationSerializer(LinkedEventsSerializer, RegistrationBaseSerializer):
+    event = JSONLDRelatedField(
+        serializer="EventSerializer",
+        many=False,
+        view_name="event-detail",
+        queryset=Event.objects.all(),
+    )
+
+    def __init__(self, instance=None, *args, **kwargs):
+        super().__init__(instance=instance, *args, **kwargs)
+
+        event_dict = kwargs["context"]["request"].data.get("event", None)
+
+        # Check user permissions if creating a new registration.
+        # LinkedEventsSerializer __init__ checks permissions in case of new registration
+        if not instance and isinstance(event_dict, dict) and "@id" in event_dict:
+            event_url = event_dict["@id"]
+            event_id = urllib.parse.unquote(event_url.rstrip("/").split("/")[-1])
+            user = kwargs["context"]["user"]
+
+            try:
+                event = Event.objects.get(pk=event_id)
+                error_message = _("User {user} cannot modify event {event}").format(
+                    user=user, event=event
+                )
+
+                if not user.is_admin(event.publisher):
+                    raise DRFPermissionDenied(error_message)
+
+                if isinstance(user, ApiKeyUser):
+                    # allow updating only if the api key matches instance data source
+                    if event.data_source != user.data_source:
+                        raise DRFPermissionDenied(_(error_message))
+                else:
+                    # allow updating only if event data_source has user_editable_resources set to True
+                    if not event.is_user_editable_resources():
+                        raise DRFPermissionDenied(_(error_message))
+            except Event.DoesNotExist:
+                pass
+
+    def create(self, request, *args, **kwargs):
+        try:
+            instance = super().create(request, *args, **kwargs)
+        except IntegrityError as error:
+            if "duplicate key value violates unique constraint" in str(error):
+                raise serializers.ValidationError(
+                    {"event": _("Event already has a registration.")}
+                )
+            else:
+                raise error
```

Comment:
You can re-raise an exception just with a plain `raise` inside an `except` block.
```suggestion
                raise
```

Reply:
Changed